modules/core/arrow-data/src/main/kotlin/arrow/data/Sum.kt

```diff
@@ -0,0 +1,52 @@
+package arrow.data
+
+import arrow.Kind
+import arrow.higherkind
+import arrow.typeclasses.Comonad
+import arrow.typeclasses.Functor
+
+@higherkind
+data class Sum<F, G, V>(
+    val left: Kind<F, V>,
+    val right: Kind<G, V>,
+    val side: Side = Side.Left
+) : SumOf<F, G, V> {
+
+  sealed class Side {
+    object Left : Side()
+    object Right : Side()
+  }
+
+  fun <A> extend(CF: Comonad<F>, CG: Comonad<G>, f: (Sum<F, G, V>) -> A): Sum<F, G, A> = CF.run {
+    CG.run {
```

Comment:
Android D8 compiler won't like this nesting, sadly. You need to put them just in the function where they're applied.

Reply:
I'm not sure if I understood this. How can I tell the compiler that F and G are Comonads then?

Reply:
Is this correct?

```kotlin
fun <A> extend(CF: Comonad<F>, CG: Comonad<G>, f: (Sum<F, G, V>) -> A): Sum<F, G, A> =
  Sum(
    CF.run { left.coflatMap { f(Sum(left, right, Side.Left)) } },
    CG.run { right.coflatMap { f(Sum(left, right, Side.Right)) } },
    side
  )  
```

Reply:
yeah, like that